diff -up zip-2.31/fileio.c.pom zip-2.31/fileio.c
--- zip-2.31/fileio.c.pom	2007-11-05 14:02:41.000000000 +0100
+++ zip-2.31/fileio.c	2007-11-05 14:03:09.000000000 +0100
@@ -976,7 +976,7 @@ lfopen(const char *path, const char *mod
     prev = mode[x];
   }
   
-  fd = open(path, flags, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+  fd = open(path, flags, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
   if (fd == -1)
     return NULL;
 
