--- zip-2.31/unix/Makefile.install	2005-02-28 08:33:50.000000000 +0100
+++ zip-2.31/unix/Makefile	2005-11-10 12:58:33.221425848 +0100
@@ -129,7 +129,7 @@
 	$(INSTALL_PROGRAM) $(ZIPS) $(BINDIR)
 	-cd $(BINDIR); $(CHMOD) $(BINFLAGS) $(ZIPS)
 	-$(INSTALL_D) $(MANDIR)
-	$(INSTALL) man/zip.1 $(MANDIR)/zip.$(manext)
+	$(INSTALL_PROGRAM) man/zip.1 $(MANDIR)/zip.$(manext)
 	$(CHMOD) $(MANFLAGS) $(MANDIR)/zip.$(manext)
 
 uninstall:
